be/src/storage/persistent_index.cpp

```diff
@@ -3738,6 +3946,85 @@ Status PersistentIndex::_merge_compaction_internal(ImmutableIndexWriter* writer,
     return Status::OK();
 }
 
+size_t PersistentIndex::_get_tmp_l1_count() {
+    return _has_l1 ? _l1_vec.size() - 1 : _l1_vec.size();
+}
+
+// There are a few steps in async compaction:
+// 1. flush l0 to l1:
+//    a. if there is only one tmp-l1 file, move this tmp-l1 to l1 file.
+//    b. if there are > 2 tmp-l1 file, then merge l0 and tmp-l1 files to new l1 file.
+//    c. if there is only one l1 file, flush l0 to new l1 file.
+//    d. if there is not l1 file exists, flush l0 to l1 file.
+// 2. move old l1 to l2. only if old l1 exist
+// 3. modify PersistentIndex meta
+Status PersistentIndex::_async_compaction(PersistentIndexMetaPB* index_meta) {
+    // 1. flush l0 to l1
+    const std::string new_l1_filename =
+            strings::Substitute("$0/index.l1.$1.$2", _path, _version.major(), _version.minor());
+    const size_t tmp_l1_cnt = _get_tmp_l1_count();
+    // maybe need to append wal in 1.a
+    bool need_append_wal = false;
+    if (tmp_l1_cnt == 1) {
+        // step 1.a
+        // move tmp l1 to l1
+        std::string tmp_l1_filename = _l1_vec[_has_l1 ? 1 : 0]->filename();
+        RETURN_IF_ERROR(FileSystem::Default()->link_file(tmp_l1_filename, new_l1_filename));
+        if (_l0->size() > 0) {
+            // check if need to append wal
+            need_append_wal = true;
+        }
+        LOG(INFO) << "PersistentIndex async compaction, link from tmp-l1: " << tmp_l1_filename
+                  << " to l1: " << new_l1_filename;
+    } else if (tmp_l1_cnt > 1) {
+        // step 1.b
+        auto writer = std::make_unique<ImmutableIndexWriter>();
+        RETURN_IF_ERROR(writer->init(new_l1_filename, _version, true));
+        // followe this rules:
+        // 1, remove delete key when l2 not exist
+        // 2. skip merge l1, only merge tmp-l1 and l0
+        RETURN_IF_ERROR(_merge_compaction_internal(writer.get(), _has_l1 ? 1 : 0, _l1_vec.size(),
+                                                   _usage_and_size_by_key_length, _l2_vec.empty() ? false : true,
+                                                   nullptr));
+        RETURN_IF_ERROR(writer->finish());
+        LOG(INFO) << "PersistentIndex async compaction, merge tmp l1, merge cnt: " << _l1_vec.size()
```

Comment:
Add the `tablet_id` in the log

Reply:
I think you can find tablet_id already in path ?

Reply:
Yes